src/future.rs

```diff
@@ -0,0 +1,71 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+struct GpuFutureInner<T> {
+    id: wgn::DeviceId,
+    result: Option<T>,
+    waker: Option<Waker>,
+}
+
+/// A Future that can poll the wgpu::Device
+pub struct GpuFuture<T> {
+    inner: Arc<Mutex<GpuFutureInner<T>>>,
+}
+
+/// A completion handle to set the result on a GpuFuture
+pub struct GpuFutureCompletion<T> {
+    inner: Arc<Mutex<GpuFutureInner<T>>>,
+}
+
+impl<T> Future for GpuFuture<T>
+where
+    T: Copy,
+{
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
+        // grab a clone of the Arc
+        let arc = Pin::into_inner(self).inner.clone();
+
+        // grab the device id and set the waker, but release the lock, so that the native callback can write to it
+        let device_id = {
+            let mut inner = arc.lock().unwrap();
+            inner.waker.replace(context.waker().clone());
+            inner.id
+        };
+
+        // polling the device should trigger the callback
+        wgn::wgpu_device_poll(device_id, true);
```

Comment:
I don't know if it's possible, but ideally we should try to find a way to do this without using `wgpu_device_poll` and instead wait for the callback to be triggered outside of `poll` somehow.

`wgpu_device_poll` and especially the `force_wait` option are only available on wgpu-native at the moment and could even be removed entirely once we figure out how the event loop/callbacks will be exposed (see https://github.com/webgpu-native/webgpu-headers/issues/18)

Reply:
It seems to be tricky. As far as I can tell, polling the future is blocking the callback from firing on the current thread (it works fine if I await the future on a second thread).

We might have to sync the future-based API down into wgpu-native to make that work.

Reply:
Yeah, I wonder if there are many examples of C callback APIs being wrapped into Rust futures out there to see how they're accomplishing it.

I'm not sure we'll be able to move many of these changes into wgpu-native, because on the web we'll have to [convert `Promise`s into `Future`s](https://rustwasm.github.io/wasm-bindgen/reference/js-promises-and-rust-futures.html) (instead of callbacks).

Reply:
If we have wgpu-native exposing an event loop, then naturally `wgpu_device_poll` will go away

Reply:
The good news for the web is that the `wasm-bindgen-futures` stuff should make converting between rust futures and JS promises pretty seamless.